tmt/utils.py

```diff
@@ -2322,3 +2331,79 @@ def find_fmf_root(path: str) -> List[str]:
         raise MetadataError(f"No fmf root present inside '{path}'.")
     fmf_roots.sort(key=lambda path: len(path))
     return fmf_roots
+
+
+# A type for callbacks given to wait()
+WaitCheckType = Callable[[], T]
+
+
+def wait(
+    parent: Common,
+    check: WaitCheckType[T],
+    timeout: datetime.timedelta,
+    tick: float = 30.0,
+    tick_increase: float = 1.0
+) -> T:
+    """
+    Wait for a condition to become true.
+
+    To test the condition state, a ``check`` callback is called every ``tick``
+    seconds until ``check`` reports a success. The callback may:
+
+    * decide the condition has been fulfilled. This is a successfull outcome,
+      ``check`` shall then simply return, and waiting ends. Or,
+    * decide more time is needed. This is not a successfull outcome, ``check``
+      shall then raise :py:clas:`WaitingIncomplete` exception, and ``wait()``
+      will try again later.
+
+    :param parent: "owner" of the wait process. Used for its logging capability.
+    :param check: a callable responsible for testing the condition. Accepts no
+        arguments. To indicate more time and attempts are needed, the callable
+        shall raise :py:class:`WaitingIncomplete`, otherwise it shall return
+        without exception. Its return value will be propagated by ``wait()`` up
+        to ``wait()``'s. All other exceptions raised by ``check`` will propagate
+        to ``wait()``'s caller as well, terminating the wait.
+    :param timeout: amount of time ``wait()`` is alowed to spend waiting for
+        successfull outcome of ``check`` call.
+    :param tick: how many seconds to wait between two consecutive calls of
+        ``check``.
+    :param tick_increase: a multiplier applied to ``tick`` after every attempt.
+    :returns: value returned by ``check`` reporting success.
+    :raises GeneralError: when ``tick`` is not a positive integer.
+    :raises WaitingTimedOutError: when time quota has been consumed.
+    """
+
+    if tick <= 0:
+        raise GeneralError('Tick must be a positive integer')
+
+    deadline = datetime.datetime.utcnow() + timeout
+
+    parent.debug(
+        'wait',
+        f"waiting for condition '{check.__name__}' with timeout {timeout}, deadline {deadline}, check every {tick} seconds")
+
+    while True:
```

Comment:
I'd suggest using `while datetime.datetime.utcnow() <= deadline` and then `raise WaitingTimedOutError()` at the end, but that's more of a personal preference, seeing `while True` is always a bit suspicious.

Reply:
I see your point, it's the endless loop, right? I don't like them too. On the other hand, I want to have `now` value, to use it not just for the test, but also for logging, and I wanted to avoid multiple calls to `utcnow()` - extra syscalls, but more importantly, different time returned every time, making logging harder to follow.

So unless there's a bug or hard concern about it, I'd stick to the current state of things.

Note: with Python 3.8, we'd have our cake and eat it too, thanks to walrus operator:

```python
while (now := datetime.datetime.utcnow()) < deadline:
  ...
raise WaitingTimedOutError()
```

Reply:
Yeah, I wish we had python 3.8 available but centos8 is a blocker :( Well, maybe one day...

Anyway, yes, no bug, lets keep it the way it is now. I will review again (hopefully next week), adding to 1.17 milestone to not forget.